File: Program_1/1(b).c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

int binary_search(int arr[],int n,int key){
    int start=0,end=n-1;
    while(start<=end){
        int mid=start+(end-start)/2;
        if(arr[mid]==key) return mid;
        else if(arr[mid]<key) start=mid+1;
        else end=mid-1;
    }
    return -1;
}

void generateRandomArray(int arr[],int n){
    for(int i=0;i<n;i++) arr[i]=rand()%10000;
}

int compare(const void* a,const void* b){
    return (*(int*)a-*(int*)b);
}

int generateRandomKey(int arr[],int n){
    int index=1+rand()%(n-1);
    return arr[index];
}


int main(){
    int sizes[]={10,50,100,500,1000,2000,3000,5000,7000,10000};
    int num_sizes=9;
    srand(time(NULL));

    for(int i=0;i<num_sizes;i++){
        int n=sizes[i];
        int* arr=(int*)malloc(n*sizeof(int));
        generateRandomArray(arr,n);
        qsort(arr,n,sizeof(int),compare);

        int key=generateRandomKey(arr,n);

        clock_t start=clock();
        for(int k=0;k<10000;k++){
            binary_search(arr,n,key);
        }
        clock_t end=clock();

        double time=((double)(end-start))/CLOCKS_PER_SEC/10000;
        printf("For input size %d time taken = %f\n",n,time);

        free(arr);
    }
    return 0;
}

File: Program_1/1(b).py
import random
import time
import matplotlib.pyplot as plt

def binary_search(arr, target):
    left = 0
    right = len(arr) - 1
    while left <= right:
        mid = (left + right) // 2
        if arr[mid] == target:
            return mid
        elif arr[mid] < target:
            left = mid + 1
        else:
            right = mid - 1
    return -1

sizes = [10, 50, 100, 500, 1000, 3000, 6000, 7000, 8000, 10000]
times = []

for n in sizes:
    arr = sorted([random.randint(0, 10000) for _ in range(n)])  # sorted for binary search
    target = random.randint(0, 10000)

    start = time.time()
    for i in range(1000):
        binary_search(arr, target)
    end = time.time()

    elapsed = (end - start) / 1000
    times.append(elapsed)
    print(f"For input size {n}, time taken = {elapsed:.6f} sec")

plt.plot(sizes, times, marker='o')
plt.xlabel("Input Size (n)")
plt.ylabel("Time Taken (sec)")
plt.title("Binary Search Performance")
plt.grid(True)
plt.show()
